lib/awesome_print/formatters/array_formatter.rb

```diff
@@ -13,60 +13,137 @@ def initialize(array, inspector)
       end
 
       def format
-        return "[]" if array == []
-
-        if array.instance_variable_defined?('@__awesome_methods__')
-          methods_array(array)
-        elsif options[:multiline]
-          width = (array.size - 1).to_s.size
-
-          data = array.inject([]) do |arr, item|
-            index = indent
-            index << colorize("[#{arr.size.to_s.rjust(width)}] ", :array) if options[:index]
-            indented do
-              arr << (index << inspector.awesome(item))
-            end
-          end
-
-          data = limited(data, width) if should_be_limited?
-          "[\n" << data.join(",\n") << "\n#{outdent}]"
+        case
+        when array.empty?
+          empty_array
+        when awesome_array?
+          methods_array
+        when multiline_array?
+          multiline_array
         else
-          "[ " << array.map{ |item| inspector.awesome(item) }.join(", ") << " ]"
+          simple_array
         end
       end
 
       private
 
-      def methods_array(a)
-        a.sort! { |x, y| x.to_s <=> y.to_s }                  # Can't simply a.sort! because of o.methods << [ :blah ]
-        object = a.instance_variable_get('@__awesome_methods__')
-        tuples = a.map do |name|
-          if name.is_a?(Symbol) || name.is_a?(String)         # Ignore garbage, ex. 42.methods << [ :blah ]
-            tuple = if object.respond_to?(name, true)         # Is this a regular method?
-              the_method = object.method(name) rescue nil     # Avoid potential ArgumentError if object#method is overridden.
-              if the_method && the_method.respond_to?(:arity) # Is this original object#method?
-                method_tuple(the_method)                      # Yes, we are good.
-              end
-            elsif object.respond_to?(:instance_method)              # Is this an unbound method?
-              method_tuple(object.instance_method(name)) rescue nil # Rescue to avoid NameError when the method is not
-            end                                                     # available (ex. File.lchmod on Ubuntu 12).
+      def awesome_array?
+        array.instance_variable_defined?('@__awesome_methods__')
+      end
+
+      def multiline_array?
+        options[:multiline]
+      end
+
+      def empty_array
+        "[]"
+      end
+
+      def simple_array
+        "[ " << array.map{ |item| inspector.awesome(item) }.join(", ") << " ]"
+      end
+
+      def multiline_array
+        data = generate_printable_array
+        data = limited(data, width(array)) if should_be_limited?
+        data = data.join(",\n")
+
+        "[\n#{data}\n#{outdent}]"
+      end
+
+      def generate_printable_array
+        array.map.with_index do |item, index|
+          array_prefix(index, width(array)).tap do |temp|
```

Comment:
Be aware that the use of `tap` pretty much kills support for Ruby 1.8.x since `tap` was introduced in 1.9. I think that's the intent, and as 1.8 cleanup we should remove `lib/awesome_print/core_ext/method.rb` and `/rails` (plugins from Rails 0.x/1.x era).